### PR TITLE
[SPARK-21228][SQL] InSet incorrect handling of structs

### PR DESCRIPTION
**Pull request by bogdanrdc**
## What changes were proposed in this pull request?
When data type is struct, InSet now uses TypeUtils.getInterpretedOrdering (similar to EqualTo) to build a TreeSet. In other cases it will use a HashSet as before (which should be faster). Similarly, In.eval uses Ordering.equiv instead of equals. 

## How was this patch tested?
New test in SQLQuerySuite.

### CODE REVIEW

**Comment by bogdanrdc** [comment]
cc @hvanhovell

**Comment by bogdanrdc** [comment]
cc @hvanhovell @cloud-fan

File: sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/expressions/predicates.scala [hunk 1]
```diff
@@ -274,24 +278,32 @@ case class InSet(child: Expression, hset: Set[Any]) extends UnaryExpression with
     }
   }
 
-  def getHSet(): Set[Any] = hset
+  @transient private[this] lazy val set = child.dataType match {
+    case _: StructType =>
```
Comment (cloud-fan) [inline sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/expressions/predicates.scala]: we should handle all complex types, which means we should also match `ArrayType` here

File: sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/expressions/predicates.scala [hunk 2]
```diff
@@ -274,24 +278,32 @@ case class InSet(child: Expression, hset: Set[Any]) extends UnaryExpression with
...
+      // for structs use interpreted ordering to be able to compare UnsafeRows with non-UnsafeRows
+      TreeSet.empty (TypeUtils.getInterpretedOrdering (child.dataType) ) ++ hset
```
Comment (cloud-fan) [inline sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/expressions/predicates.scala]: nit: no space near `()` please.

File: sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/expressions/predicates.scala [hunk 3]
```diff
@@ -274,24 +278,32 @@ case class InSet(child: Expression, hset: Set[Any]) extends UnaryExpression with
...
-    val hsetTerm = ctx.freshName("hset")
+    val setTerm = ctx.freshName("set")
     val hasNullTerm = ctx.freshName("hasNull")
-    ctx.addMutableState(setName, hsetTerm,
-      s"$hsetTerm = (($InSetName)references[${ctx.references.size - 1}]).getHSet();")
-    ctx.addMutableState("boolean", hasNullTerm, s"$hasNullTerm = $hsetTerm.contains(null);")
+    ctx.addMutableState(setName, setTerm,
+      s"$setTerm = (($InSetName)references[${ctx.references.size - 1}]).getSet();")
+    ctx.addMutableState("boolean", hasNullTerm,
+      s"$hasNullTerm = ${if (hasNull) "true" else "false"};")
```
Comment (cloud-fan) [inline sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/expressions/predicates.scala]: this means `hasNullTerm` is a constant at compile time, we can just use `true` or `false` in the generated code.

File: sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/expressions/predicates.scala [hunk 4]
```diff
@@ -274,24 +278,32 @@ case class InSet(child: Expression, hset: Set[Any]) extends UnaryExpression with
...
     ev.copy(code = s"""
       ${childGen.code}
       boolean ${ev.isNull} = ${childGen.isNull};
       boolean ${ev.value} = false;
       if (!${ev.isNull}) {
-        ${ev.value} = $hsetTerm.contains(${childGen.value});
+        ${ev.value} = $setTerm.contains(${childGen.value});
         if (!${ev.value} && $hasNullTerm) {
```
Comment (cloud-fan) [inline sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/expressions/predicates.scala]:
```
val setNull = if (hasNull) {
  s"""
    if (!${ev.value}) {
      ${ev.isNull} = true;
    }
  """
} else {
  ""
}
```
Then we can  just write `$setNull` here.

File: sql/core/src/test/scala/org/apache/spark/sql/SQLQuerySuite.scala [hunk 5]
```diff
@@ -2616,4 +2616,27 @@ class SQLQuerySuite extends QueryTest with SharedSQLContext {
...
+          """
+            |SELECT * from
+            | (SELECT MIN(a) as minA FROM A) AA -- this Aggregate will return UnsafeRows
+            | -- the IN will become InSet with a Set of GenericInternalRows
+            | -- a GenericInternalRow is never equal to an UnsafeRow so the query would
+            | -- returns 0 results, which is incorrect
+            | WHERE minA IN (NAMED_STRUCT('a', 1L, 'b', 1L), NAMED_STRUCT('a', 2L, 'b', 2L),
+            |   NAMED_STRUCT('a', 3L, 'b', 3L))
+          """.stripMargin)
+        df.explain
```
Comment (viirya) [inline sql/core/src/test/scala/org/apache/spark/sql/SQLQuerySuite.scala]: Do we need explain here?

File: sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/expressions/predicates.scala (hunk 1 above)
Comment (viirya) [inline sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/expressions/predicates.scala]: Should we check if all fields are orderable?

Comment (viirya) [inline sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/expressions/predicates.scala]: This approach may not work for the structs including the types that can't be ordered.

Comment (bogdanrdc) [inline sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/expressions/predicates.scala]: I checked and indeed it is needed for ArrayType too. MapType is not comparable.
I think dealing with uncomparable types should be different issue. It wouldn't work before either. In.doCodeGen would fail with IllegalArgumentException. EqualTo has a check specifically for Map to throw AnalysisException. That could be a fix. Or implement comparing for Maps.

File: sql/catalyst/src/test/scala/org/apache/spark/sql/catalyst/expressions/PredicateSuite.scala [hunk 6]
```diff
@@ -35,7 +35,8 @@ class PredicateSuite extends SparkFunSuite with ExpressionEvalHelper {
     test(s"3VL $name") {
       truthTable.foreach {
         case (l, r, answer) =>
-          val expr = op(NonFoldableLiteral(l, BooleanType), NonFoldableLiteral(r, BooleanType))
+          val expr = op(NonFoldableLiteral.create(l, BooleanType),
```
Comment (cloud-fan) [inline sql/catalyst/src/test/scala/org/apache/spark/sql/catalyst/expressions/PredicateSuite.scala]: why these changes? i.e. adding `.create`

File: sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/expressions/predicates.scala [hunk 7]
```diff
@@ -274,27 +278,40 @@ case class InSet(child: Expression, hset: Set[Any]) extends UnaryExpression with
     }
   }
 
-  def getHSet(): Set[Any] = hset
+  @transient private[this] lazy val set = child.dataType match {
+    case _: AtomicType => hset
+    case _: NullType => hset
+    case _ =>
+      // for structs use interpreted ordering to be able to compare UnsafeRows with non-UnsafeRows
+      TreeSet.empty(TypeUtils.getInterpretedOrdering (child.dataType) ) ++ hset
```
Comment (cloud-fan) [inline sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/expressions/predicates.scala]: no space near `()` please, `.getInterpretedOrdering(child.dataType))`

**Comment by cloud-fan** [comment]
LGTM, let's resolve the conflicts

File: sql/catalyst/src/test/scala/org/apache/spark/sql/catalyst/expressions/PredicateSuite.scala (hunk 6 above)
Comment (bogdanrdc) [inline sql/catalyst/src/test/scala/org/apache/spark/sql/catalyst/expressions/PredicateSuite.scala]: The problem is that NonFoldableLiteral(java.math.BigDecimal, DecimalType) is keeping the java value while NonFoldableLiteral.create(_,_) does Literal.create which transforms java.math.BigDecimal to Decimal (catalyst type).
It's a bit of weird code. Literal has the same problem. Literal(_,_) is not the same as Literal.create(_,_). So, one should always use .create and never the 2 parameter constructor.
Otherwise you get java values in the execution, which in my case was failing because there is no ordering for non catalyst types.

**Comment by cloud-fan** [comment]
thanks, merging to master!
this conflicts with branch 2.2, can you send a new PR? thanks!
